include/cuda/std/memory_resource

```diff
@@ -0,0 +1,266 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include "cstddef"
+#include "detail/__config"
+#include "detail/__pragma_push"
+#include "stream"
+#include "version"
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/**
+ * @brief Specifies the kind of memory of an allocation.
+ *
+ * A memory allocation's kind determines where the memory can be accessed from
+ * and the performance characteristics of the accesses.
+ *
+ */
+enum class memory_kind {
+  device,  ///< Device memory accessible only from device
+  unified, ///< Unified memory accessible from both host and device
+  pinned,  ///< Page-locked system memory accessible from both host and device
+  host     ///< System memory only accessible from host code
+};
+
+/**
+ * @brief Abstract interface for synchronous memory allocation.
+ *
+ * @tparam _Kind The `memory_kind` of the allocated memory.
+ */
+template <memory_kind _Kind> class memory_resource {
+public:
+  static constexpr memory_kind memory_kind = _Kind;
```

Comment:
This kind of code causes compiler-dependent problems in derived classes - namely, the symbol `memory_kind` in a derived class can be resolved to the enum name or to the field name. Older GCC preferred to interpret it as the field, whereas newer GCC preferred to interpret is as the type.

Reply:
Rename the field to `allocated_memory_kind`?

Reply:
Is `kind` too ambiguous for the member name? `memory_resource::kind`?

Reply:
I don't think so, as long as the type name is unambiguous (`memory_kind`).